data/functions.sql

```diff
@@ -640,21 +640,94 @@ END;
 $$ LANGUAGE plpgsql IMMUTABLE;
 
 
--- calculate the list of refs (or names) of transit routes that
--- a particular station is part of. this ends up being a very
--- complex function because it's bouncing around between the
--- nodes, ways and relations to calculate membership of various
--- sets.
+-- for the purposes of sweeping for related public transit features,
+-- "interesting" tags are ones which define relations which group public
+-- transport features, or group features together in a "site". these are:
+--   * public_transport = stop_area
+--   * public_transport = stop_area_group
+--   * type = stop_area
+--   * type = stop_area_group
+--   * type = site
+--
+CREATE OR REPLACE FUNCTION mz_is_interesting_transit_relation(
+  tags hstore)
+RETURNS BOOLEAN AS $$
+BEGIN
+  RETURN
+    tags->'public_transport' IN ('stop_area', 'stop_area_group') OR
+    tags->'type' IN ('stop_area', 'stop_area_group', 'site');
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+-- extract a name for a transit route relation. this can expand comma
+-- separated lists and prefers to use the ref rather than the name.
+CREATE OR REPLACE FUNCTION mz_transit_route_name(
+  tags hstore)
+RETURNS text AS $$
+DECLARE
+  route_name text;
+BEGIN
+  SELECT TRIM(BOTH FROM "name") INTO route_name
+    FROM (
+      SELECT UNNEST(string_to_array(COALESCE(
+        -- prefer ref as it's less likely to contain the destination name
+        tags->'ref',
+        tags->'name'
+      ), ',')) AS "name"
+    ) refs_and_names
+    LIMIT 1;
+  RETURN route_name;
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+DO $$
+BEGIN
+  IF EXISTS (SELECT 1 FROM pg_type WHERE typname = 'mz_transit_routes_rettype') THEN
+    DROP TYPE mz_transit_routes_rettype CASCADE;
+  END IF;
+
+  CREATE TYPE mz_transit_routes_rettype AS (
+    -- the "top-most" site or public_transport relation found while exploring
+    -- the graph of objects around the station.
+    root_relation_id BIGINT,
+
+    -- the "score" of the station. this takes the number of rail routes, the
+    -- sum of the number of subway and light rail routes, and the sum of the
+    -- number of tram and railway routes. if it is an interchange (i.e: both
+    -- rail and subway / light_rail routes) then doubling both those numbers.
+    -- these then get maxed out at 9, and turned into a 3-digit number.
+    score SMALLINT,
+
+    -- the list of distinct refs (or names if refs aren't present) on each
+    -- type of route.
+    train_routes      text[],
+    subway_routes     text[],
+    light_rail_routes text[],
+    tram_routes       text[],
+    railway_routes    text[]
+  );
+END$$;
```

Comment:
Yeah, especially with the `WITH RECURSIVE` CTEs below... that was a whole new thing, and full of interesting restrictions.